contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,303 @@
+        uint32 timeElapsed = uint32(block.timestamp) - _tradingStartTimes[token];
+        // get the current price by adjusting the eth amount with the exp decay formula
+        price.ethAmount = uint128(ExpDecayMath.calcExpDecay(price.ethAmount, timeElapsed, _priceDecayHalfLife));
+        // return the price
+        return price;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function trade(
+        Token token,
+        uint128 amount
+    ) external payable nonReentrant validToken(token) greaterThanZero(amount) {
+        uint128 ethRequired = expectedTradeInput(token, amount);
+        // revert if trade requires 0 eth
+        if (ethRequired == 0) {
+            revert InvalidTrade();
+        }
+        // check enough eth has been sent for the trade
+        if (msg.value < ethRequired) {
+            revert InsufficientNativeTokenSent();
+        }
+        // transfer the tokens to caller
+        token.safeTransfer(msg.sender, amount);
+
+        // refund any excess eth to caller
+        if (msg.value > ethRequired) {
+            payable(msg.sender).sendValue(msg.value - ethRequired);
+        }
+
+        // emit event
+        emit TokenTraded(msg.sender, token, amount, ethRequired);
+    }
+
+    function _setMarketPriceMultiply(uint32 newMarketPriceMultiply) private {
+        uint32 prevMarketPriceMultiply = _marketPriceMultiply;
+
+        // return if the market price multiply is the same
+        if (prevMarketPriceMultiply == newMarketPriceMultiply) {
+            return;
+        }
+
+        _marketPriceMultiply = newMarketPriceMultiply;
+
+        emit MarketPriceMultiplyUpdated(prevMarketPriceMultiply, newMarketPriceMultiply);
+    }
+
+    function _setPriceDecayHalfLife(uint32 newPriceDecayHalfLife) private {
+        uint32 prevPriceDecayHalfLife = _priceDecayHalfLife;
+
+        // return if the price decay half-life is the same
+        if (prevPriceDecayHalfLife == newPriceDecayHalfLife) {
+            return;
+        }
+
+        _priceDecayHalfLife = newPriceDecayHalfLife;
+
+        emit PriceDecayHalfLifeUpdated(prevPriceDecayHalfLife, newPriceDecayHalfLife);
+    }
+
+    function _validToken(Token token) private view {
+        // validate token is not the native token
+        if (token.isNative()) {
+            revert InvalidToken();
+        }
+        // validate trading is enabled for token
+        if (!_tradingEnabled(token)) {
+            revert TradingDisabled();
+        }
+    }
+
+    function _validPrice(Price memory price) private pure {
+        if (price.tokenAmount == 0 || price.ethAmount == 0) {
+            revert InvalidPrice();
+        }
+    }
+
+    /**
+     * @dev return true if trading is enabled for token
+     */
+    function _tradingEnabled(Token token) private view returns (bool) {
+        return _tradingStartTimes[token] != 0;
+    }
+
+    function uncheckedInc(uint256 i) private pure returns (uint256 j) {
```

Comment:
Nitpick: can be moved to something like `Utils`

Reply:
Nitpick #2: not being used by the contract

Reply:
Removed